### PR TITLE
feat: Ticket pre-selected for 1 type of tickets

### LINKED ISSUE #5449: Ticket Order: If there is only one ticket type, please pre-select '1' instead of '0'
If there is only one ticket type, please pre-select 1 instead of 0. If there are more ticket types please leave the implementation as is for this case.

Compare https://eventyay.com/e/1f720482
Screenshot from 2020-11-02 12-11-10

### PR DESCRIPTION
**Pull request by daretobedifferent18**
Fixes #5449 

#### Short description of what this resolves:
If there is only one ticket type , pre-select ticket with minOrder.

#### Changes proposed in this pull request:

- For one type of ticket, minOrder of tickets is pre-selected
- For minOrder of tickets pre-selected, Order now button is also working,
- Also total and taxes are calulated before only showing that ticket is pre-selected.

#### BEFORE (no pre-select)
Screenshot from 2020-11-12 20-20-45

#### AFTER (pre-selected)
(for minOrder=0)
Screenshot from 2020-11-12 20-24-42
(for minOrder=3)
Screenshot from 2020-11-12 20-28-54

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia).
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The acceptance, integration, unit tests and linter pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

### CODE REVIEW

**Comment by daretobedifferent18** [comment]
@iamareebjamal Please see the changes and implementation I have made.

File: app/components/public/ticket-list.js [hunk 1]
```diff
@@ -230,6 +230,15 @@ export default Component.extend(FormMixin, {
...
     }
+    if (this.tickets.length === 1) {
+      let preSelect = this.tickets[0].minOrder;
+      if (!this.tickets[0].minOrder) {
+        preSelect = 1;
+      }
+      this.tickets[0].set('orderQuantity', preSelect);
+      this.order.tickets.addObject(this.tickets[0]);
+      this.send('updateOrderAmount');
+    }
```
Comment (iamareebjamal) [inline app/components/public/ticket-list.js:241]: This should be above togglePromotionalCode

File: app/components/public/ticket-list.js [hunk 2]
```diff
@@ -230,6 +230,15 @@ export default Component.extend(FormMixin, {
     if (this.code) {
       this.send('togglePromotionalCode', this.code);
     }
+    if (this.tickets.length === 1) {
+      let preSelect = this.tickets[0].minOrder;
+      if (!this.tickets[0].minOrder) {
+        preSelect = 1;
+      }
```
Comment (iamareebjamal) [inline app/components/public/ticket-list.js:237]:
```suggestion
      const preSelect = Math.max(this.tickets[0].minOrder, 1);
```

File: app/components/public/ticket-list.js [hunk 3]
```diff
@@ -99,6 +99,21 @@ export default Component.extend(FormMixin, {
...
+    if (this.code) {
+      this.send('togglePromotionalCode', this.code);
+    }
+    if (this.tickets.length === 1) {
+      const preSelect = Math.max(this.tickets[0].minOrder, 1);
+      this.tickets[0].set('orderQuantity', preSelect);
+      this.order.tickets.addObject(this.tickets[0]);
+      this.send('updateOrderAmount');
+    }
+  },
```
Comment (iamareebjamal) [inline app/components/public/ticket-list.js:115]: Why did you move it up?

Comment (daretobedifferent18) [reply app/components/public/ticket-list.js:115]: U didn' told me to move this code above togglePromotionalCode?

Comment (iamareebjamal) [reply app/components/public/ticket-list.js:115]: Above `this.send('togglePromotionalCode')`

File: app/templates/components/public/ticket-list.hbs [hunk 4]
```diff
@@ -88,12 +88,19 @@
                     @id={{concat ticket.id "_quantity"}}
                     @value={{ticket.orderQuantity}} />
                   <i class="dropdown icon"></i>
-                  <div class="default text">0</div>
+                  <div class="default text">{{if (eq this.tickets.length 1) (if (gt ticket.minOrder 1) ticket.minOrder 1) 0}}</div>
```
Comment (iamareebjamal) [inline app/templates/components/public/ticket-list.hbs:91]: Why the need of this logic now?

File: app/templates/components/public/ticket-list.hbs [hunk 5]
```diff
@@ -88,12 +88,19 @@
...
                   <div class="menu">
+                  {{#if (eq this.tickets.length 1)}}
+                    <div class="item" data-value="{{map-value mapper ticket.minOrder}}">{{ticket.minOrder}}</div>
+                    {{#each (range (add ticket.minOrder 1) (add ticket.maxOrder 1)) as |count|}}
+                      <div class="item" data-value="{{map-value mapper count}}">{{count}}</div>
+                    {{/each}}
+                  {{else}}
```
Comment (iamareebjamal) [inline app/templates/components/public/ticket-list.hbs:98]: This logic shouldn't be needed

File: app/templates/components/public/ticket-list.hbs (hunk 4 above)
Comment (daretobedifferent18) [reply app/templates/components/public/ticket-list.hbs:91]: bcz we need to show if no of ticketes type is '1' and minOrer is 0 then it should be 1 else minOrder 
we can improve this by 
```handlebars
{{if (eq this.tickets.length 1) (max ticket.minOrder 1) 0}}
```

**Comment by daretobedifferent18** [comment]
I have also improved the bug which was coming when minOrder is 0 if no of types of tickets is more then 1
so it solves for very situation

BUG 
Screenshot from 2020-11-13 13-40-54
showing 0 two times (fixed now)

File: app/templates/components/public/ticket-list.hbs [hunk 6]
```diff
@@ -88,10 +88,10 @@
                     @id={{concat ticket.id "_quantity"}}
                     @value={{ticket.orderQuantity}} />
                   <i class="dropdown icon"></i>
-                  <div class="default text">0</div>
+                  <div class="default text">{{if (eq this.tickets.length 1) (max ticket.minOrder 1) 0}}</div>
```
Comment (iamareebjamal) [inline app/templates/components/public/ticket-list.hbs:91]: Wouldn't `{{ticker.orderQuantity}}` work here?

Comment (daretobedifferent18) [reply app/templates/components/public/ticket-list.hbs:91]: no bcz the value of ticket.orderQuantity is undefined intially and meantime the dom is rendered completely after that ticket.orderQuantity value is changed.

Comment (daretobedifferent18) [reply app/templates/components/public/ticket-list.hbs:91]: but I think it should be working but its not. it is showing blank

Comment (iamareebjamal) [reply app/templates/components/public/ticket-list.hbs:91]: `{{or ticket.orderQuantity 0}}`

Comment (daretobedifferent18) [reply app/templates/components/public/ticket-list.hbs:91]: this will work OK for more then one type of tickets but in case of only 1 type of tickets still ticket.orderQuanitity is undefined so it is 0 there also.
and I don't know why in this it is wokring correct
```handlebars
 <Input
    @type="hidden"
    @id={{concat ticket.id "_quantity"}}
    @value={{ticket.orderQuantity}} />
```
but not here
```handlebars
 <div class="default text">{{or ticket.orderQuanitity 0}}</div>
```

**Review by iamareebjamal — APPROVED** [review]
